### DISCUSSION TITLE
Atomicity of operation?

### DISCUSSION THREAD

**Discussion by phuongnd08** [dragonflydb/dragonfly#772] [Q&A] [answered]
Let say I need to get the value of a key, increase it by x, and set the new value to database. With redis lua script, I can make sure if I send N operations in parallel, they are executed without race condition because redis is single threaded. How can I achive this with dragonfly?

**Comment by romange** [comment, marked as answer]
You can do it exactly the same in Dragonfly.
It preserves the atomicity of lua scripts. 

You do not have to be single threaded to guarantee atomicity.

**Reply by romange** [reply]
Btw you are the first farmer i know that asked me about lua scripting and atomicity. 🏅

**Reply by phuongnd08** [reply]
The reason I asked because if two racing thread is getting a value, then change them, in parallel, the final result depends on which one finish last. So the only way to guarantee no racing is that one thread should not start until the other thread finish running (aka single threading). Wonder how dragonfly handle that?

**Reply by romange** [reply]
"single threading" is only one way to achieve atomicity. Another way could be using locks: once a thread starts, it locks the keys it touches and does not release them until they are processed.  We do not use locks; we use [shared-nothing](https://github.com/dragonflydb/dragonfly/blob/main/docs/df-share-nothing.md) architecture together with [state of the art](https://www.cs.umd.edu/~abadi/papers/vldbj-vll.pdf)  algorithm from 2015 that employs logical locks to schedule operations.

underneath, the algorithm tracks which keys are touched by which operations and orders them in such a way that it won't break atomicity guarantees.
